### ISSUE TITLE
onRestoreInstanceState() should be called before onResume()

### ISSUE THREAD

**Issue by yankeppey** [konmik/nucleus#101]
Hey, thanks for the library.

Several times I caught the exception from here:

```
    /**
     * {@link android.app.Activity#onCreate(Bundle)}, {@link android.app.Fragment#onCreate(Bundle)}, {@link android.view.View#onRestoreInstanceState(Parcelable)}.
     */
    public void onRestoreInstanceState(Bundle presenterState) {
        if (presenter != null)
            throw new IllegalArgumentException("onRestoreInstanceState() should be called before onResume()");
        this.bundle = ParcelFn.unmarshall(ParcelFn.marshall(presenterState));
    }
```

The exception message seems too low-level for any library user. After some source code exploring, I found, that we can't call `getPresenter()` in any point before `onResume()`. Thus, for instance, method references like
> private SomeView.ClickListener mViewClickListener = getPresenter()::viewClicked;

can crash the app, because getPresenter() is obviously called before `onCreate`.

Also I caught the crash several times in other callbacks, probably, in `FragmentStatePagerAdapter` (which can call onRestoreInstanceState before or even without onCreate, sorry, I don't remember a lot of details)

Is it reasonable to:
1. Replace throwing exception to some warning message, leaving developer's responsibility to save/restore state?
2. (more complex) to separate Presenter constructor and onCreate() methods, so that some time, before onCreate call, Presenter can exist without state restored?

**Comment by konmik** [comment]
You CAN call `getPresenter()` before `onResume()`. You can't call `onRestoreInstanceState` before `onResume()`.

`getPresenter()` does not work before parent's `onCreate` call, this is obvious. Nothing in Android works before parent's `onCreate`.

1. I can't imagine a use case when such behaviour (restoration of presenter after view attachment) will be reasonable.
2. Are You asking to add another lifecycle? The purpose of all this stuff were to have *less* lifecycles. Try to manage using as less lifecycle events as possible, this makes reliable code. Lifecycles is one of the biggest mistakes in Android design.

**Comment by gautamjain** [comment]
Hi @konmik 

I've been struggling with this exception for a while.  Here's a simple project that replicates the exception:  https://github.com/gautamjain/nucleus-test

The first fragment, `AlphaFragment`, has a _child_ called `SimpleCard`.  These extend from `NucleusSupportFragment` and `NucleusLayout` respectively.  They both have their own presenters.

I then _replace_ `AlphaFragment` with another fragment, `BetaFragment` (and add the transaction to the back stack).   Clicking the back button at this point throws the exception.

Here's a simplified log of what happens: 

```
MainActivity: ⇢ onBackPressed()

AlphaFragment: ⇢ onCreateView()
    SimpleCard: ⇢ <init>()
    SimpleCard: ⇠ <init> [4ms]
AlphaFragment: ⇠ onCreateView [39ms] = android.widget.LinearLayout

SimpleCard: ⇢ onAttachedToWindow()
  SimpleCardPresenter: ⇢ onCreate(savedState=null)
  SimpleCardPresenter: ⇠ onCreate [0ms]
  SimpleCardPresenter: ⇢ onTakeView(view=com.example.nucleusstatetest.SimpleCard)
  SimpleCardPresenter: ⇠ onTakeView [0ms]
SimpleCard: ⇠ onAttachedToWindow [3ms]

SimpleCard: ⇢ onRestoreInstanceState(
                      state=Bundle[{presenter_state=Bundle[{presenter=Bundle[{nucleus.presenter.RxPresenter#requested=[]}], presenter_id=SimpleCardPresenter], parent_state=android.view.AbsSavedState}])

InputEventSender: Exception dispatching finished signal.
MessageQueue-JNI: Exception in MessageQueue callback: handleReceiveCallback
MessageQueue-JNI: java.lang.IllegalArgumentException: onRestoreInstanceState() should be called before onResume()
                                                                                    at nucleus.view.PresenterLifecycleDelegate.onRestoreInstanceState(PresenterLifecycleDelegate.java:84)
```

How could such a situation be avoided?  `FragmentManager` is responsible for initiating the state restoration of the fragment's child views.

**Comment by konmik** [comment]
Hi @gautamjain ,

I never worked with child fragments, but afaik, state restoration should be done before attachment to window. Is this a usual behavior for child fragments?

**Comment by gautamjain** [comment]
It's actually not a child fragment.  `SimpleCard` extends from `NucleusLayout`.   In this case, the child `SimpleCard` gets created, attached, and then view state restored.

**Comment by sgc-code** [comment]
I'm having the same issue here with onRestoreInstanceState() being called after onAttachedToWindow() for views.
I've been trying to avoid NuclearLayout in favor of nucleus fragments but it will be great to have Views with controllers as well

**Comment by konmik** [comment]
It looks to me that child fragments violate view lifecycle somehow.

Do you have any ideas how to handle it?

I do not see reliable solutions to overcome this.

**Comment by konmik** [comment]
Oops, I understood! Child fragments is another story, sorry! 

Need some investigation.

**Comment by sgc-code** [comment]
I don't have a solution for this. Maybe we could defer the Presenter.onCreate() to be called afterthe state is restored, but we don't know if there's any state to restore.
We could try to go up the hierarchy to find the fragment containing the view, and then check if we have some state to restore.

Or maybe this is just a bug in the support library that we can report and wait for a fix

**Comment by konmik** [comment]
What if we will just ignore `onRestoreInstanceState` that are fired after `onAttachedToWindow`?

**Comment by sgc-code** [comment]
Ignoring the `onRestoreInstanceState` would make the saved state lost.
Also, we need the Bundle for creating the presenter, otherwise we will create a new presenter after rotation instead of reusing the old one. So we can't create the presenter without Bundle with the current architecture.

**Comment by konmik** [comment]
@sgc-code You're right, FragmentManager creates a new View instance, its state (including presenter) must be restored.

I can also try to delete the presenter that was initialized during `onAttachedToWindow` and create a new one from the saved state. This will introduce some redundancy in the case of the lifecycle bug, but will still work.

**Comment by sgc-code** [comment]
I think you can't do that because you will lose the state changes that happen between `onAttachedToWindow` and `onRestoreInstanceState`

What about delaying the creation of the presenter to the `onRestoreInstanceState` method?

**Comment by konmik** [comment]
@sgc-code *normally* `onResoreInstanceState` is called after `onAttachedToWindow`. If we will just wait for `onRestoreInstanceState`, most of the cases we will wait forever, because we can't know if we should wait for it or start working.

**Comment by sgc-code** [comment]
We can change a flag on the View so we know if `onRestoreInstanceState` was already called and we can proceed to presenter creation in that case.

The tricky part would be the first time the view is created in the nested fragment situation, because the onRestoreInstanceState won't be called as there's no state to restore. I think maybe  it's possible to workaround this problem by querying the container fragment for its state

**Comment by konmik** [comment]
My final recommendation on this topic will be to not use `NucleusView` inside of fragments. Fragments violate view lifecycle, so I see no way to implement stateful presenters reliably here.
